### PR TITLE
add an optional KvStorage in form of Remote LMDB [PR#21] module for TCSListener and EnclaveManager

### PR DESCRIPTION
**Pull request by sammyne**
1. bootstrap the LMDB as an independent service in scripts/tcs_startup.sh if `LMDBHelperProxy` is employed
    2. add `LMDBHelperProxy` as an remote option to interact with LMDB for TCSListener and EnclaveManager

Signed-off-by: Sammy <lixiangmin01@baidu.com>

### CODE REVIEW

**Comment by m3ngyang** [comment]
#21

File: scripts/tcs_startup.sh [hunk 1]
```diff
@@ -19,8 +19,18 @@ listener="${TCF_HOME}/examples/common/python/connectors/direct/tcs_listener/tcs_
 # Read Listener port from config file
 listener_port=`grep listener_port ${TCF_HOME}/config/tcs_config.toml | awk {'print $3'}`
 
+lmdb_server="${TCF_HOME}/examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py"
+# the following port must be synchronized to that of KvStorage.database_url 
+# in config/tcs_config.toml
+LMDB_SERVER_PORT=9090
```
Comment (manju956) [inline scripts/tcs_startup.sh]: Can you read the LMDB_SERVER_PORT from config file similar to listener_port?

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+ # mypy
+.mypy_cache/
+.dmypy.json
+dmypy.json
+
+ # Pyre type checker
+.pyre/
+
+tmpdir/
```
Comment (manju956) [inline .gitignore]: Insert line at the end of file.

**Comment by manojgop** [comment]
Could you update PR comment to reflect what it's doing instead of "bring PR#21 into action"

File: .gitignore (hunk 2 above)
Comment (sammyne) [reply .gitignore]: update cd711736a8b2b953501e3fe18a4a2680bbb496a6

File: scripts/tcs_startup.sh (hunk 1 above)
Comment (sammyne) [reply scripts/tcs_startup.sh]: done cd711736a8b2b953501e3fe18a4a2680bbb496a6

File: examples/enclave_manager/tcf_enclave_manager/enclave_manager.py [hunk 3]
```diff
@@ -225,6 +230,8 @@ def create_enclave_signup_data():
     return enclave_signup_data
 
 # -----------------------------------------------------------------
+
+
 def execute_work_order(enclave_data, input_json_str, indent=4):
     """
     Submits workorder reqeust to Worker enclave and retrieves the response
```
Comment (danintel) [inline examples/enclave_manager/tcf_enclave_manager/enclave_manager.py]: s/reqeust/request/

Comment (sammyne) [reply examples/enclave_manager/tcf_enclave_manager/enclave_manager.py]: Sorry, I don't get it. Would you please explain your point in detail? @danintel

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,126 @@
+# Byte-compiled / optimized / DLL files
+__pycache__/
+*.py[cod]
+*$py.class
+
+ # C extensions
+*.so
+
+ # Distribution / packaging
+.Python
```
Comment (manojgop) [inline .gitignore]: Are all these file generated today. Could we add only those files that needs to be ignored by git.

File: config/tcs_config.toml [hunk 5]
```diff
@@ -14,6 +14,7 @@
 
 [KvStorage]
 StoragePath = "config/Kv_Shared_tmp"
+database_url = "http://localhost:9090"
```
Comment (manojgop) [inline config/tcs_config.toml]: Can we add a database mode direct/remote. Based  on the mode the implementation can use either KvStorage directly or remote KvStorage.

File: examples/enclave_manager/tcf_enclave_manager/enclave_manager.py (hunk 3 above)
Comment (manojgop) [reply examples/enclave_manager/tcf_enclave_manager/enclave_manager.py]: I guess he mean to fix spelling errors. Search and Replace reqeust/request/.

File: .gitignore (hunk 4 above)
Comment (sammyne) [reply .gitignore]: I just copy that from [here](https://github.com/github/gitignore/blob/master/Python.gitignore), which hosts gitignore for various programming languages, and has been reviewed by nice people. Picking out those for this project doesn't worth it, personally speaking.

File: config/tcs_config.toml (hunk 5 above)
Comment (sammyne) [reply config/tcs_config.toml]: done as 484fbbe36d0b4c2d970fdcf7fe7a0a30d1a2d5f9

File: tc/sgx/common/packages/db_store/lmdb_store.cpp [hunk 6]
```diff
@@ -466,4 +466,4 @@ std::string tcf::db_store::db_store_get_all(
 #endif
 
     return table_keys;
-}
+}
```
Comment (manojgop) [inline tc/sgx/common/packages/db_store/lmdb_store.cpp]: Please add new line at end of file

File: tools/clean.sh [hunk 7]
```diff
@@ -42,9 +42,15 @@ make clean
 cd $SRCDIR/tc/sgx/trusted_worker_manager/enclave
 rm -rf build deps
 
+cd $SRCDIR/examples/enclave_manager/tcf_enclave_manager
+rm -rf tcf_enclave.py tcf_enclave_wrap.cpp
```
Comment (manojgop) [inline tools/clean.sh]: These files are removed in clean section in Makefile in examples/enclave_manager. I think its better to invoke make clean

File: tools/clean.sh [hunk 8]
```diff
@@ -42,9 +42,15 @@ make clean
...
+
 cd $SRCDIR/tc/sgx/common
 rm -rf build
 
+## wrapper
+rm -f $SRCDIR/tc/sgx/trusted_worker_manager/enclave_wrapper/*_u.*
+
```
Comment (manojgop) [inline tools/clean.sh]: the clean section in Makefile in examples/enclave_manager removes these files.

File: scripts/tcs_startup.sh [hunk 9]
```diff
@@ -19,8 +19,15 @@ listener="${TCF_HOME}/examples/common/python/connectors/direct/tcs_listener/tcs_
...
 listener_port=`grep listener_port ${TCF_HOME}/config/tcs_config.toml | awk {'print $3'}`
 
+lmdb_server="${TCF_HOME}/examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py"
+
 start_tcs_components()
 {
+    echo "start the lmdb server"
+    python3 ${lmdb_server} &
+    sleep 3s
+    echo "lmdb server started"
```
Comment (manojgop) [inline scripts/tcs_startup.sh]: Can we create lmdb remote server functionality if lmdb remote_url is set in config file ? or we can use a argument to startup script ?

File: examples/enclave_manager/tcf_enclave_manager/enclave_manager.py (hunk 3 above)
Comment (sammyne) [reply examples/enclave_manager/tcf_enclave_manager/enclave_manager.py]: Gotcha ~

File: scripts/tcs_startup.sh (hunk 9 above)
Comment (sammyne) [reply scripts/tcs_startup.sh]: The startup of lmdb remote server has been handled in the `lmdb_listener.py`. If `remote_url` is set in config file, this server would simply quit with warning and the local LMDB would be employed.

File: tools/clean.sh (hunk 7 above)
Comment (sammyne) [reply tools/clean.sh]: done as 1a0efa538e05b07b64280f3afd7a13a1cc58ad06

File: tools/clean.sh (hunk 8 above)
Comment (sammyne) [reply tools/clean.sh]: done as 1a0efa538e05b07b64280f3afd7a13a1cc58ad06

File: tc/sgx/common/packages/db_store/lmdb_store.cpp (hunk 6 above)
Comment (sammyne) [reply tc/sgx/common/packages/db_store/lmdb_store.cpp]: done as e91d6a89dd498d293d8295b983d4c0612af5a5ce

Comment (sammyne) [reply tc/sgx/common/packages/db_store/lmdb_store.cpp]: done as e91d6a89dd498d293d8295b983d4c0612af5a5ce

File: scripts/tcs_startup.sh (hunk 9 above)
Comment (sammyne) [reply scripts/tcs_startup.sh]: implement as 3fa74144d8679cf83ec5694355266d3b7650b666, which would employ the remote LMDB if `remote_url` exists in config/tcs_config.toml

File: tools/clean.sh [hunk 10]
```diff
@@ -42,6 +42,9 @@ make clean
 cd $SRCDIR/tc/sgx/trusted_worker_manager/enclave
 rm -rf build deps
 
+cd $SRCDIR/examples/enclave_manager
+SGX_MODE=SIM make clean
```
Comment (manojgop) [inline tools/clean.sh]: SGX_MODE is set as an environmental variable and shouldn't be hard coded here. Also if the changes in clean.sh is unrelated of this PR , can you raise a separate PR for clean.sh changes. Preferred approach is to keep unrelated changes in different PR.

Comment (sammyne) [reply tools/clean.sh]: This hardcoded env setting is to address the error out logic as follows 

https://github.com/hyperledger-labs/trusted-compute-framework/blob/ab50a1368d60a36626ac171b1a72a37152ec2b5a/examples/enclave_manager/Makefile#L18-L20

If missing, the `clean` target for this folder would be broken.

Forgive me for this growing PR =_= 
Separate this change out takes effort due to too many commits now. So sincerely ask for leaving it this time, please.

I would never raise PR so large. Thanks for the good advice 😄

Comment (sammyne) [reply tools/clean.sh]: Separation is done as https://github.com/hyperledger-labs/trusted-compute-framework/pull/42

Comment (sammyne) [reply tools/clean.sh]: Separation goes this PR https://github.com/hyperledger-labs/trusted-compute-framework/pull/42

File: tools/clean.sh (hunk 7 above)
Comment (sammyne) [reply tools/clean.sh]: this work has been moved into separate PR https://github.com/hyperledger-labs/trusted-compute-framework/pull/42

File: tools/clean.sh (hunk 8 above)
Comment (sammyne) [reply tools/clean.sh]: this work has been moved into separate PR https://github.com/hyperledger-labs/trusted-compute-framework/pull/42

**Comment by sammyne** [comment]
> Please add signed-off field to all of your commits in PR. Jenkins job which executes sanity tests is failing because of missing field.

All commits had been appended the signed-off field ~ @manju956 
Is it possible the sanity testing scripts has bugs?

**Comment by manju956** [comment]
> > Please add signed-off field to all of your commits in PR. Jenkins job which executes sanity tests is failing because of missing field.
> 
> All commits had been appended the signed-off field ~ @manju956
> I just don't know why the tests failed = =

this PR has conflicts. Can you please resolve and push changes? then we can see why jenkins is failing.

File: .gitignore [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+.pytest_cache/
+
+ # Translations
+*.mo
+*.pot
+
+ # Django stuff:
+*.log
+local_settings.py
+db.sqlite3
```
Comment (Ram-srini) [inline .gitignore]: Please remove whatever is not required, for example db.sqlite3, Transactions etc, which are not used in our project. It is goof to keep what is required for our project.

Comment (sammyne) [reply .gitignore]: I have rebase my work on the latest master, and remote this .gitignore, see 25f67189141a4d943436e9c2c03b843981766aeb @Ram-srini

**Comment by sammyne** [comment]
> > > Please add signed-off field to all of your commits in PR. Jenkins job which executes sanity tests is failing because of missing field.
> > 
> > 
> > All commits had been appended the signed-off field ~ @manju956
> > I just don't know why the tests failed = =
> 
> this PR has conflicts. Can you please resolve and push changes? then we can see why jenkins is failing.
> Resolve conflict in file -> examples/enclave_manager/tcf_enclave_manager/enclave_manager.py

I have rebase my work on the latest master, see 25f67189141a4d943436e9c2c03b843981766aeb. This should resolve all conflicts @manju956

**Comment by manju956** [comment]
> > > > Please add signed-off field to all of your commits in PR. Jenkins job which executes sanity tests is failing because of missing field.
> > > 
> > > 
> > > All commits had been appended the signed-off field ~ @manju956
> > > I just don't know why the tests failed = =
> > 
> > 
> > this PR has conflicts. Can you please resolve and push changes? then we can see why jenkins is failing.
> > Resolve conflict in file -> examples/enclave_manager/tcf_enclave_manager/enclave_manager.py
> 
> I have rebase my work on the latest master, see [25f6718](https://github.com/hyperledger-labs/trusted-compute-framework/commit/25f67189141a4d943436e9c2c03b843981766aeb). This should resolve all conflicts @manju956

Thanks, after rebase sanity tests in jenkins has passed.

File: examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py [hunk 12]
```diff
@@ -27,92 +28,105 @@
...
 
 # -----------------------------------------------------------------
-def parse_command_line(config, args) :
+
+
+def parse_command_line(config, args):
 
     global bind_port
 
+    if config.get("KvStorage") is None or config["KvStorage"].get("remote_url") is None:
```
Comment (manojgop) [inline examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py]: you meant "and" instead of "or" ?

if (config.get("KvStorage") is None) and (config["KvStorage"].get("remote_url") is None):

Comment (sammyne) [reply examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py]: `or` not `and`：either `KvStorage` or `KvStorage.remote_url` shouldn't employ the remote LMDB @manojgop

**Comment by sammyne** [comment]
> Could you update PR comment to reflect what it's doing instead of "bring PR#21 into action"

already done ~

File: examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py (hunk 12 above)
Comment (manojgop) [reply examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py]: either StoragePath or remote_url should be present. If both are None then its an error condition

if (config.get("KvStorage").get("StoragePath) is None) and (config["KvStorage"].get("remote_url") is None):

Comment (sammyne) [reply examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py]: Yes, it should error out in case of both being None.

For `if (config.get("KvStorage").get("StoragePath") is None) and (config["KvStorage"].get("remote_url") is None)`, 
the chaining call `config.get("KvStorage").get("StoragePath")` would throw in case of `config.get('KvStorage')` being `None`~

For ` if config.get("KvStorage") is None or config["KvStorage"].get("remote_url") is None`,
if `config.get("KvStorage") is None` is `true`, we can early abort since `config["KvStorage"].get("remote_url")` mustn't exist then. Otherwise (i.e., `config["KvStorage"]` isn't `None`), we still need to ensure that `remote_url` goes under it. Only then should we employ the remote LMDB safely without crashing the program ~ @manojgop

Comment (manojgop) [reply examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py]: What I meant is if both StoragePath and remote_url is not set i.e if both are None then it's an error condition.  If "KvStorage" is none then it's an error condition.

Comment (m3ngyang) [reply examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py]: config loader should have the ability to ensure that the given configuration is verified and rightful before passing to other modules, instead of checking parameter everywhere. Maybe we can make a TODO in the `config/config.py`.

Comment (sammyne) [reply examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py]: I got it. 

However, if the `and` check is taken,  we would miss out the error case with `KvStorage.StoragePath` set and `KvStorage.remote_url` unset, which should quit the listener due to no suitable url to serve on.

But `lmdb_listener` should only care whether `KvStorage.remote_url` is set, and leave the checking against `KvStorage.StoragePath` to the `LMDBRequestHandler`.

Comment (manojgop) [reply examples/common/python/shared_kv/remote_lmdb/lmdb_listener.py]: ok.
